lib/yjit_metrics/analysis.rb

```diff
@@ -0,0 +1,179 @@
+
+module YJITMetrics
+  module Analysis
+    REPORT_LINK = "[Analysis Report](https://speed.yjit.org/analysis.txt)"
+
+    # Build report by reading files from provided dir.
+    def self.report_from_dir(dir, benchmarks: nil)
+      metrics = self.metrics
+      # Load results from the last #{count} benchmark runs.
+      count = metrics.map(&:count).max
+      # We only need to load the files for the following configs ("yjit_stats"...).
+      configs = metrics.map(&:config).uniq
+
+      # data = {yjit_stats: {"x86_64_yjit_stats" => [result_hash, ...], ...}
+      data = configs.each_with_object({}) do |config, h|
+        YJITMetrics::PLATFORMS.each do |platform|
+          files = Dir.glob("**/*_basic_benchmark_#{platform}_#{config}.json", base: dir).sort.last(count).map { |f| File.join(dir, f) }
+
+          runs = files.map { |f| JSON.parse(File.read(f)) }.group_by { |x| x["ruby_config_name"] }
+          # Append data to h[:yjit_stats]["x86_64_yjit_stats"].
+          (h[config] ||= {}).merge!(runs) { |k, oldv, newv| oldv + newv }
+        end
+      end
+
+      report_from_data(data, metrics:, benchmarks:)
+    end
+
+    # Build report from hash of data (built by `report_from_dir`).
+    def self.report_from_data(data, metrics: self.metrics, benchmarks: nil)
+      # {ratio_in_yjit: {"x86_64_yjit_stats" => {"benchmark_name" => results_of_check, ...}}
+      metrics.each_with_object({}) do |metric, h|
+        data[metric.config].each_pair do |config_name, run|
+          metric.check(run, benchmarks:).then do |result|
+            # h[:ratio_in_yjit]["x86_64_yjit_stats"] = ...
+            (h[metric.name] ||= {})[config_name] = result unless result.empty?
+          end
+        end
+      end.then do |results|
+        Report.new(results)
+      end
+    end
+
+    def self.metrics
+      Metric.subclasses.map(&:new)
+    end
+
+    class Report
+      attr_reader :results
+
+      def initialize(results)
+        @results = results
+      end
+
+      # Format regression data into a notification message.
+      def regression_notification
+        msg = Hash.new { |h, k| h[k] = [] }
+
+        results.each_pair do |metric, h|
+          h.each_pair do |platform, values|
+            values.sort.each do |benchmark, report|
+              next unless regression = report[:regression]
+
+              msg["#{metric} #{platform}"] << "- `#{benchmark}` regression: #{regression}"
+            end
+          end
+        end
+
+        return if msg.empty?
+
+        lines = msg.sort.flatten
+        lines << REPORT_LINK
+        lines.join("\n")
+      end
+    end
+
+    class Metric
+      include Stats
+
+      def config
+        self.class::CONFIG
+      end
+
+      def count
+        self.class::COUNT
+      end
+
+      def name
+        self.class::NAME
+      end
+    end
+
+    class RatioInYJIT < Metric
+      CONFIG = :yjit_stats
+      COUNT = 30
+      NAME = :ratio_in_yjit
+      ROUND = 2
+      TOLERANCE = 0.1
```

Comment:
Can you comment on what is the unit of the tolerance?
What is count?
What is round?